### PR TITLE
[ADD]Added route field in sale type, Add onchange it

### PR DESCRIPTION
**Pull request by chetanrdhaduk**
Task URL:
https://www.druidoo.io/web#id=2137&action=367&model=project.task&view_type=form&menu_id=237

### CODE REVIEW

File: sale_order_type/README.rst [hunk 1]
```diff
@@ -77,6 +79,7 @@ Authors
 * Tecnativa
 * Agile Business Group
 * Niboo
+* Druidoo
```
Comment (ivantodorovich) [inline sale_order_type/README.rst]: Don't add Druidoo as author

File: sale_order_type/tests/test_sale_order_type.py [hunk 2]
```diff
@@ -92,3 +160,12 @@ def test_invoice_change_partner(self):
...
         self.assertEqual(invoice.sale_type_id, self.sale_type)
+
+    def test_sale_order_flow_route(self):
+        order = self.sale_order_model.create(self.get_sale_order_vals())
+        order.type_id = self.sale_type_1.id
+        order.onchange_type_id()
+        self.assertEqual(order.type_id.route_id, order.order_line[0].route_id)
+        order.type_id = self.sale_type_2.id
+        order.onchange_type_id()
+        self.assertEqual(order.type_id.route_id, order.order_line[0].route_id)
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
```

I think we can remove this `sale_type_2` from the tests.
It's not testing anything new, right? It's the same flow than `sale_type_1`, or am I missing something?

Comment (ivantodorovich) [reply sale_order_type/tests/test_sale_order_type.py]: We should probably test adding a product to the sale order, though.
We expect the sale order line to have the same route defined in the sale order type

File: sale_order_type/tests/test_sale_order_type.py [hunk 3]
```diff
@@ -49,6 +49,74 @@ def setUp(self):
...
+                            "name": "SO-2 -> Customer",
+                            "action": "pull",
+                            "picking_type_id": self.ref("stock.picking_type_in"),
+                            "location_src_id": self.ref("stock.stock_location_14"),
+                            "location_id": self.ref("stock.stock_location_customers"),
+                        },
+                    )
+                ],
+            }
+        )
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
```

File: sale_order_type/tests/test_sale_order_type.py [hunk 4]
```diff
@@ -49,6 +49,74 @@ def setUp(self):
...
+                "sequence_id": self.sequence.id,
+                "journal_id": self.journal.id,
+                "warehouse_id": self.warehouse.id,
+                "picking_policy": "one",
+                "payment_term_id": self.immediate_payment.id,
+                "pricelist_id": self.sale_pricelist.id,
+                "incoterm_id": self.free_carrier.id,
+                "route_id": self.route_2.id,
+            }
+        )
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
```

File: sale_order_type/tests/test_sale_order_type.py [hunk 5]
```diff
@@ -49,6 +49,74 @@ def setUp(self):
             }
         )
         self.partner.sale_type = self.sale_type
+        self.route_1 = self.env["stock.location.route"].create(
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
        self.sale_route = self.env["stock.location.route"].create(
```

File: sale_order_type/tests/test_sale_order_type.py [hunk 6]
```diff
@@ -49,6 +49,74 @@ def setUp(self):
...
+            {
+                "name": "Test Sale Order Type-1",
+                "sequence_id": self.sequence.id,
+                "journal_id": self.journal.id,
+                "warehouse_id": self.warehouse.id,
+                "picking_policy": "one",
+                "payment_term_id": self.immediate_payment.id,
+                "pricelist_id": self.sale_pricelist.id,
+                "incoterm_id": self.free_carrier.id,
+                "route_id": self.route_1.id,
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
                "route_id": self.sale_route.id,
```

File: sale_order_type/tests/test_sale_order_type.py [hunk 7]
```diff
@@ -49,6 +49,74 @@ def setUp(self):
...
+                            "action": "pull",
+                            "picking_type_id": self.ref("stock.picking_type_in"),
+                            "location_src_id": self.ref("stock.stock_location_14"),
+                            "location_id": self.ref("stock.stock_location_customers"),
+                        },
+                    )
+                ],
+            }
+        )
+        self.sale_type_1 = self.sale_type_model.create(
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
        self.sale_type_route = self.sale_type_model.create(
```

File: sale_order_type/tests/test_sale_order_type.py [hunk 8]
```diff
@@ -92,3 +160,12 @@ def test_invoice_change_partner(self):
         self.assertEqual(invoice.sale_type_id, self.sale_type)
         invoice = self.invoice_model.new({"partner_id": self.partner_child_1.id})
         self.assertEqual(invoice.sale_type_id, self.sale_type)
+
+    def test_sale_order_flow_route(self):
+        order = self.sale_order_model.create(self.get_sale_order_vals())
+        order.type_id = self.sale_type_1.id
```
Comment (ivantodorovich) [inline sale_order_type/tests/test_sale_order_type.py]:
```suggestion
        order.type_id = self.sale_type_route.id
```

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
A few comments. After you finish it, please squash all your commits into a single one and we're ready to push to OCA

File: sale_order_type/models/sale.py [hunk 9]
```diff
@@ -67,3 +69,14 @@ def _prepare_invoice(self):
...
             res["sale_type_id"] = self.type_id.id
         return res
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    @api.onchange("product_id")
+    def product_id_change(self):
+        domain = super(SaleOrderLine, self).product_id_change()
```
Comment (ivantodorovich) [inline sale_order_type/models/sale.py]:
```suggestion
        res = super(SaleOrderLine, self).product_id_change()
```

File: sale_order_type/models/sale.py [hunk 10]
```diff
@@ -67,3 +69,14 @@ def _prepare_invoice(self):
...
+        if self.order_id.type_id.route_id:
+            self.update({"route_id": self.order_id.type_id.route_id})
+        return domain
```
Comment (ivantodorovich) [inline sale_order_type/models/sale.py]:
```suggestion
        return res
```

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
Also I've encountered two issues:

1) Can we remove the route_id of the lines if the sale_order_type doesn't have one?
I'm trying to keep it consistent when changing sale order types. If the sale order type doesn't have a route, I think we should remove it.

2) I've encountered this issue when changing the sale order type of a sale order that already has products. Can you look into it? 

It seems it only happens when the order is not saved yet. If the order is saved, then there's no issue. It's weird.

image

Please try to find a solution. Maybe it's because of the use of `write` inside the onchange?

----

Screen record from 2020-04-22 16 46 02

In the video you can see the 2 issues I've just described

File: sale_order_type/views/res_partner_view.xml [hunk 11]
```diff
@@ -4,12 +4,12 @@
...
-                    <separator string="Sales Order Type" colspan="2" />
-                    <field name="sale_type" />
-                </group>
-            </page>
+            <xpath
+                expr="//page[@name='sales_purchases']/group/group[@name='sale']"
+                position="inside"
+            >
+                <field name="sale_type" />
+            </xpath>
```
Comment (ivantodorovich) [inline sale_order_type/views/res_partner_view.xml]:
```suggestion
            <group name="sale" position="inside">
                <field name="sale_type" />
            </group>
```

Doesn't it work like this?
If it does, it's much cleaner

File: sale_order_type/models/sale.py [hunk 12]
```diff
@@ -51,6 +51,10 @@ def onchange_type_id(self):
                 vals.update({"incoterm": order_type.incoterm_id})
             if vals:
                 order.update(vals)
+            if order_type.route_id:
+                order.order_line.write({"route_id": order_type.route_id.id})
+            else:
+                order.order_line.write({"route_id": False})
```
Comment (ivantodorovich) [inline sale_order_type/models/sale.py]:
```suggestion
            order.order_line.write({"route_id": order_type.route_id.id})
```

This is basically the same in only 1 line of code.
Can you check, though, about this errors that appears when changing the order type if the order is not saved yet?
I left a picture and video in one of the comments, I think it's still not resolved

**Review by ivantodorovich — CHANGES_REQUESTED** [review]

File: sale_order_type/models/sale.py (hunk 12 above)
Comment (chetanrdhaduk) [reply sale_order_type/models/sale.py]: I checked the picture and video & tested multiple times, But I can't see the same error.
It's working properly.

**Comment by ivantodorovich** [comment]
Closing to push to OCA

**Pull request closed without merging**